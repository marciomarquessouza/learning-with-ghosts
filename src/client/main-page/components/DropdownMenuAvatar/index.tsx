import React, { Fragment } from 'react'
import { Menu, Transition } from '@headlessui/react'

import Avatar from '../Avatar'

export interface DropdownMenuAvatarProps {
    name: string
    email: string | null
    photoUrl?: string | null
    onLogout: () => void
}

function DropdownMenuAvatar({ name, email, photoUrl, onLogout }: DropdownMenuAvatarProps) {
    const handleLogout = (e: React.MouseEvent<HTMLElement>) => {
        e.preventDefault()
        onLogout()
    }

    return (
        <Menu as="div" className="relative inline-block text-left">
            <div>
                <Menu.Button className="">
                    <Avatar name={name} photoUrl={photoUrl} />
                </Menu.Button>
            </div>
            <Transition
                as={Fragment}
                enter="transition ease-out duration-100"
                enterFrom="transform opacity-0 scale-95"
                enterTo="transform opacity-100 scale-100"
                leave="transition ease-in duration-75"
                leaveFrom="transform opacity-100 scale-100"
                leaveTo="transform opacity-0 scale-95"
            >
                <Menu.Items className="absolute right-0 z-10 mt-2 w-56 origin-top-right rounded-md bg-white shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none">
                    <div className="py-1">
                        <Menu.Item>
                            {() => (
                                <button
                                    type="button"
                                    onClick={handleLogout}
                                    className="
                                        bg-gray-100 text-gray-900 block w-full px-4 py-2 text-left text-sm"
                                >
                                    Sign out
                                </button>
                            )}
                        </Menu.Item>
                    </div>
                </Menu.Items>
            </Transition>
        </Menu>
    )
}

export default DropdownMenuAvatar
