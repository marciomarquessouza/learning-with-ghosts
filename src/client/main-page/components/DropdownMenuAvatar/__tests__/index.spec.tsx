import React from 'react'

import { render, fireEvent, waitFor } from '@testing-library/react'

import DropdownMenuAvatar from '..'

describe('components >> DropdownMenuAvatar', () => {
    const defaultMockedProps = {
        name: '__MOCK__NAME__',
        email: '__MOCK__EMAIL__',
        onLogout: jest.fn(),
    }

    it('should render the DropdownMenuAvatar component properly', () => {
        const { container } = render(<DropdownMenuAvatar {...defaultMockedProps} />)

        expect(container).toMatchSnapshot('DropdownMenuAvatar-default')
    })

    it('should fire the logout event when the Sign Out button is pressed', async () => {
        const mockOnLogout = jest.fn()
        const mockedProps = {
            ...defaultMockedProps,
            onLogout: mockOnLogout,
        }
        const { getByText } = render(<DropdownMenuAvatar {...mockedProps} />)

        fireEvent.click(getByText('MN'))

        await waitFor(() => {
            fireEvent.click(getByText('Sign out'))
        })

        expect(mockOnLogout).toHaveBeenCalledTimes(1)
    })
})
