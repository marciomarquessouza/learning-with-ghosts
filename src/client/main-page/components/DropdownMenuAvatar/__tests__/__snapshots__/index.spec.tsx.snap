// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components >> DropdownMenuAvatar should render the DropdownMenuAvatar component properly: DropdownMenuAvatar-default 1`] = `
<div>
  <div
    class="relative inline-block text-left"
  >
    <div>
      <button
        aria-expanded="false"
        aria-haspopup="true"
        class=""
        id="headlessui-menu-button-:r0:"
        type="button"
      >
        <div>
          <span
            class="inline-flex h-12 w-12 items-center justify-center rounded-full bg-gray-500"
          >
            <span
              class="text-lg font-medium leading-none text-white"
            >
              MN
            </span>
          </span>
        </div>
      </button>
    </div>
  </div>
</div>
`;
